File: Source/Fastore.Client/ClientException.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.Serialization;
using System.Text;

namespace Alphora.Fastore.Client
{
	public class ClientException : Exception
	{
		public ClientException() { }
		public ClientException(string message) : base(message) { }
		public ClientException(string message, Codes code) : base(message) 
		{
			Data.Add("Code", code);
		}
		protected ClientException(SerializationInfo info, StreamingContext context) : base(info, context) { }
		public ClientException(string message, Exception innerException) : base(message, innerException) { }

		public enum Codes
		{
			/// <summary> There is no worker for the present column. </summary>
			NoWorkerForColumn = 10000,
            NoWorkersInHive = 10001

			// TODO: fill out rest of codes and update throw sites
		}

		public static void ThrowErrors(List<Exception> errors)
		{
			if (errors.Count > 0)
			{
				if (errors.Count == 1)
					throw errors[0];
				else
					throw new AggregateException(errors);
			}
		}

		public static void ForceCleanup(params Action[] actions)
		{
			var errors = new List<Exception>();
			foreach (var action in actions)
				try
				{
					action();
				}
				catch (Exception e)
				{
					errors.Add(e);
				}
			ThrowErrors(errors);
		}
	}
}

File: Source/FastoreClient/ClientException.cpp
#include "ClientException.h"
#include <functional>
#include <vector>

#include "../FastoreCore/Log/Syslog.h"

using namespace fastore::client;
using namespace std;

using fastore::Log;
using fastore::log_endl;
using fastore::log_info;
using fastore::log_err;

ClientException::ClientException() 
    : std::runtime_error(std::string()), Code(Codes::General)
{
}

ClientException::ClientException(const std::string &message) 
  : std::runtime_error(message), Code(Codes::General)
{
}

ClientException::ClientException(const std::string &message, Codes code) 
  : std::runtime_error(message), Code(code)
{
}

ClientException::ClientException(const std::string &message, std::exception &inner) 
  : std::runtime_error(message), Code(Codes::General), Inner(inner)
{
}

void ClientException::ThrowErrors(std::vector<std::exception> &errors)
{
	if (errors.size() > 0)
	{
		// TODO: implement aggregate exceptions
		Log << log_info << "ClientException::" << __func__ 
			<< ": throwing 1 of " << errors.size() << " errors" << log_endl;
		throw errors[0];
	}
}

void ClientException::ForceCleanup(vector<function<void()>> statements)
{
	auto errors = vector<exception>();
	for (function<void()> statement : statements)
	{
		try
		{
			statement();
		}
		catch (std::exception &e)
		{
			errors.push_back(e);
		}
	}
	ThrowErrors(errors);
}
